### ISSUE TITLE
【topk / radix-select】PTO ISA 与当前编译器 TileOP API 一致性对比（需修改 4 / 一致 1；含完整 top-K 抽取相的前置 API 缺口）

### ISSUE THREAD

**Issue by liujiang833** [PTO-ISA/SuperNPUBench#15]
> 关联 #14（同一口径的 API 一致性核对，本 issue 聚焦 `kernels/sort/topk`）。

## 目的

以 `kernels/sort/topk.hpp` 用到的 tile-level 指令为基准，逐条比对当前编译器
`tileop_api.hpp` 的 C++ 接口与 PTO ISA 文档定义，**识别 API 层面的不一致并推动编译器修改**。

**范围**：只看编译器对外 **API 签名层面**的一致性（名称、参数、行为语义、类型支持、变体缺失）。
**编译器内部如何实现（内联汇编 / SIMT / 块指令 / 模型支持）不作为判定依据。**

**PTO ISA 权威来源**：`github.com/hw-native-sys/pto-isa@main:include/pto/common/pto_instr.hpp`
（文档站 `pto-isa.github.io/.../thistogram` 为 stub，不含签名，勿作来源）。

---

## A. shipped 路径（topk 实际编译进 kernel 的指令）

| # | 类别 | PTO ISA | 当前编译器 API | 需要的修改 |
|---|---|---|---|---|
| A1 | **需要修改** | `THISTOGRAM` | `THISTOGRAM(dst, src, Idx, int ByteId)`（在 `jcore/template_asm.hpp`） | ① 字节选择 `int ByteId` → 模板枚举 `HistByte byte`（编译期）；② `src`/`idx` 拆成独立类型参数（现强制 `src` 与 `Idx` 同类型，无法表达 src=u16 + idx=u8）；③ 补 `WaitEvents...` + 返回 `RecordEvent`；④ 从裸汇编 helper 提升为标准 `tileop_api.hpp` 一等接口 |
| A2 | | `TLOAD` | `TCOPYIN(dst, src)` | 改名 `TCOPYIN` → `TLOAD`（参数/行为已一致） |
| A3 | | `TSTORE` | `TCOPYOUT(dst, src)` | 改名 `TCOPYOUT` → `TSTORE`；补 `AtomicType` / `preQuantScalar` 参数 |
| A4 | | `TEXPANDS` | `TEXPANDSCALAR(dst, s)` | 改名 `TEXPANDSCALAR` → `TEXPANDS`（参数/行为已一致） |
| A5 | **完全一致** | `TADD` | `TADD(dst, src0, src1)` | 无差异 |

**统计（shipped 路径）**：需要修改 4 · 完全一致 1。

### THISTOGRAM 签名对比

PTO ISA 权威签名（`pto_instr.hpp:2360`）：
```cpp
template <HistByte byte, typename TileDataDst, typename TileDataSrc, typename TileDataIdx,
          typename... WaitEvents>
PTO_INST RecordEvent THISTOGRAM(TileDataDst &dst, TileDataSrc &src, TileDataIdx &idx,
                                WaitEvents &... events);
// 用法：THISTOGRAM<HistByte::BYTE_1>(dst, src, idx);   （src=u16 tile、idx=u8 tile）
```
当前编译器（`jcore/template_asm.hpp:198`）：
```cpp
template <is_tile_data_v tile_shape_out, is_tile_data_v tile_shape_in>
void THISTOGRAM(tile_shape_out &dst, tile_shape_in &src, tile_shape_in &Idx, int ByteId);
```

| 维度 | PTO ISA | 当前编译器 | 需修改 |
|---|---|---|---|
| 字节选择 | 模板枚举 `HistByte byte` | 运行值 `int ByteId` | ✔ |
| idx 类型 | 独立 `TileDataIdx` | 与 src 共用 `tile_shape_in` | ✔ |
| 事件 | `WaitEvents...` + `RecordEvent`（内部 `TSYNC(events...)`） | 无、返回 `void` | ✔ |
| 操作数顺序/语义 | `(dst 256-bin 前缀累积, src 值, idx 前缀过滤)` | 一致 | — |
| 暴露位置 | `pto_instr.hpp` 一等 intrinsic | 仅 `jcore/template_asm.hpp` 裸汇编 helper | ✔ |

> 注：PTO `THISTOGRAM` **无** `tmp` / `PrecisionType` 参数。

---

## B. 贯穿性不一致（所有指令共通，最高影响面）

**补齐 `WaitEvents...` / `RecordEvent` 事件模型**：PTO 每条 intrinsic 内部首句都是
`TSYNC(events...)`，是强制约定；当前编译器 API 全线缺失（不带、返回 `void`）。
建议作为一次统一改造，而非逐条打补丁。

**命名映射**（需编译器改名以对齐 PTO）：
`TCOPYIN→TLOAD`、`TCOPYOUT→TSTORE`、`TEXPANDSCALAR→TEXPANDS`、`TSELECT→TSEL`（与 #14 #17–#20 一致）。

---

## C. masked-select 抽取相（**完整 top-K 的前置依赖**）

当前 `topk` 只算到 threshold（第 K 大元素的“值”）并 O(1) 校验，**未做真正的元素抽取**
（`masked_select` 是未调用的标量参考代码）。要把 topk 补成完整 top-K，就需要
`masked_select + gather + 局部排序` 相，而其高效 tile-op 化卡在下列 API 缺口：

| # | 类别 | PTO ISA | 当前编译器 API | 需要的修改 |
|---|---|---|---|---|
| C1 | **需要修改** | `TCVT` | `TCVT(dst, src)` | 补 `RoundMode` / `SaturationMode` / `tmp`（同 #14 #10） |
| C2 | | `TCMP` + `TCMPS` | `TCMP(dst, src0, src1, cmpMode)` | 补 `WaitEvents`；**`EQ` 模式当前写死只支持 `int32_t`**，需放宽到含 u16 等；**新增 tile-标量比较变体 `TCMPS`** |
| C3 | | `TSEL` | `TSELECT(dst, cond, src0, src1)` | 改名 `TSELECT` → `TSEL`；补 `tmp`（PTO `TSEL` 用 packed predicate，需 tmp 解包缓冲；当前用 unpacked 掩码 tile）；补 `WaitEvents` |

**现状说明**：旧硬阻塞（`TCAST` u16→u32 崩 ClockHands pass）已修复（TCAST 改走 TCVT 块指令）。
现在 tile-op 抽取“可行但很丑/低效”——需 `TCAST u16→int32`（绕 EQ 的 int32 限制）+
`TEXPANDS 广播标量`（绕无 TCMPS）+ `TCMP` 出掩码。补齐 C 档后即可干净、原生地 tile-op 化。

---

## 参考

- 本仓库对照文档：`kernels/sort/TOPK_PTOISA_COMPLETENESS.md`
- 当前编译器 API：`Linx-TileOP-API/include/common/tileop_api.hpp`、`jcore/template_asm.hpp`
- PTO ISA 权威源：`hw-native-sys/pto-isa@main` — `include/pto/common/pto_instr.hpp`
  （`THISTOGRAM`:2360）、用法样例 `tests/npu/a5/src/st/testcase/thistogram/thistogram_kernel.cpp`

**Comment by zhoubot** [comment]
## RFC update: hierarchical pair TopK with `TPARTVIEW` / `TASSEMBLY`

The current upstream `pypto-lib` implementation provides a second, and likely better, completion direction than the radix-select path originally tracked by this issue.

Exact upstream evidence inspected at `hw-native-sys/pypto-lib@7fd3bafcf1d80f52ef6d75ce36c67aec40c215c3`:

- [`examples/advanced/topk.py`](https://github.com/hw-native-sys/pypto-lib/blob/7fd3bafcf1d80f52ef6d75ce36c67aec40c215c3/examples/advanced/topk.py): `sort32(values, indices)` -> `mrgsort(64)` -> `mrgsort(256)` -> first K interleaved pairs -> P0101/P1010 gather.
- [`models/qwen3_14b/topk_select.py`](https://github.com/hw-native-sys/pypto-lib/blob/7fd3bafcf1d80f52ef6d75ce36c67aec40c215c3/models/qwen3_14b/topk_select.py): each 2048-wide group keeps local Top-32, group candidates are staged into value/id pools padded to 4096, and a second sort/merge selects global Top-32. Final candidate-pool positions are mapped through `candidate_ids` to original token IDs.

This route avoids the variable-length threshold-bucket compaction problem. Each group produces two naturally aligned fragments:

```text
32 FP32 values = 128 B
32 INT32 ids   = 128 B
```

Those fragments map directly to fixed TileArray slots and complete `B.ASSEMBLE` coverage.

### Proposed Linx TileArray sentinel

```cpp
#include <common/pto_tileop.hpp>
#include <cstdint>
#include <utility>

using namespace pto;

constexpr int VocabSize = 65536;
constexpr int GroupWidth = 2048;
constexpr int TopK = 32;
constexpr int NumGroups = VocabSize / GroupWidth;
constexpr int CandidateCount = NumGroups * TopK;

using LogitsGlobal = global_tensor<float, RowMajor<1, VocabSize>>;

using LogitsParentShape = Tile<Location::Vec, float, 1, VocabSize, BLayout::RowMajor>;
using LogitsParent = SharedTile<LogitsParentShape>;

using GroupValueTile = Tile<Location::Vec, float, 1, GroupWidth, BLayout::RowMajor>;
using GroupValueView = SubTileView<LogitsParent, GroupValueTile>;
using LogitsGroupArray = BorrowedTileArray<LogitsParent, GroupValueTile, 1, NumGroups>;
using GroupIndexTile = Tile<Location::Vec, std::uint32_t, 1, GroupWidth, BLayout::RowMajor>;

using GroupPairTile = SortPairTile<float, std::uint32_t, GroupWidth>;
using GroupTopPairTile = SortPairTile<float, std::uint32_t, TopK>;
using GroupTopPairView = SubTileView<GroupPairTile, GroupTopPairTile>;
using GroupTopPairArray = BorrowedTileArray<GroupPairTile, GroupTopPairTile, 1, GroupWidth / TopK>;

using CandidateValueSubTile = Tile<Location::Vec, float, 1, TopK, BLayout::RowMajor>;
using CandidateIndexSubTile = Tile<Location::Vec, std::int32_t, 1, TopK, BLayout::RowMajor>;
using CandidateValueArray = TileArray<CandidateValueSubTile, 1, NumGroups>;
using CandidateIndexArray = TileArray<CandidateIndexSubTile, 1, NumGroups>;
using CandidateValueSlot = TileArrayOutputRef<CandidateValueSubTile>;
using CandidateIndexSlot = TileArrayOutputRef<CandidateIndexSubTile>;

using CandidateValueParent = Tile<Location::Vec, float, 1, CandidateCount, BLayout::RowMajor>;
using CandidateIndexParent = Tile<Location::Vec, std::int32_t, 1, CandidateCount, BLayout::RowMajor>;
using CandidatePositionTile = Tile<Location::Vec, std::uint32_t, 1, CandidateCount, BLayout::RowMajor>;
using CandidatePairTile = SortPairTile<float, std::uint32_t, CandidateCount>;

using FinalTopPairTile = SortPairTile<float, std::uint32_t, TopK>;
using FinalTopPairView = SubTileView<CandidatePairTile, FinalTopPairTile>;
using FinalTopPairArray = BorrowedTileArray<CandidatePairTile, FinalTopPairTile, 1, CandidateCount / TopK>;
using FinalTopKValueTile = Tile<Location::Vec, float, 1, TopK, BLayout::RowMajor>;
using FinalTopKPositionTile = Tile<Location::Vec, std::uint32_t, 1, TopK, BLayout::RowMajor>;
using FinalTopKIndexTile = Tile<Location::Vec, std::int32_t, 1, TopK, BLayout::RowMajor>;

struct TopKResult {
    FinalTopKValueTile Values;
    FinalTopKIndexTile Indices;
};

TopKResult topk_select(LogitsGlobal &logits_global)
{
    LogitsParent logits_parent;
    TLOAD(logits_parent, logits_global);

    LogitsGroupArray logits_groups = TPARTVIEW<GroupValueTile, 1, NumGroups>(logits_parent);
    CandidateValueArray candidate_value_array;
    CandidateIndexArray candidate_index_array;

#pragma clang loop unroll(full)
    for (int group = 0; group < NumGroups; ++group) {
        GroupValueView group_values = logits_groups[0][group];
        GroupIndexTile group_indices;
        GroupPairTile sorted32_pairs;
        GroupPairTile merged64_pairs;
        GroupPairTile merged256_pairs;
        GroupPairTile merged1024_pairs;

        TCI(group_indices, group * GroupWidth);
        TSORT32(sorted32_pairs, group_values, group_indices);
        TMRGSORT<64>(merged64_pairs, sorted32_pairs);
        TMRGSORT<256>(merged256_pairs, merged64_pairs);
        TMRGSORT<1024>(merged1024_pairs, merged256_pairs);

        GroupTopPairArray group_pair_blocks = TPARTVIEW<GroupTopPairTile, 1, GroupWidth / TopK>(merged1024_pairs);
        GroupTopPairView group_top_pairs = group_pair_blocks[0][0];
        CandidateValueSlot candidate_value_slot = candidate_value_array[0][group];
        CandidateIndexSlot candidate_index_slot = candidate_index_array[0][group];

        TGATHER(candidate_value_slot, group_top_pairs, MaskPattern::P0101);
        TGATHER(candidate_index_slot, group_top_pairs, MaskPattern::P1010);
    }

    CandidateValueParent candidate_values = TASSEMBLY<CandidateValueParent>(std::move(candidate_value_array));
    CandidateIndexParent candidate_indices = TASSEMBLY<CandidateIndexParent>(std::move(candidate_index_array));

    CandidatePositionTile candidate_positions;
    CandidatePairTile candidate_sorted32;
    CandidatePairTile candidate_merged64;
    CandidatePairTile candidate_merged256;
    CandidatePairTile candidate_merged1024;

    TCI(candidate_positions, 0);
    TSORT32(candidate_sorted32, candidate_values, candidate_positions);
    TMRGSORT<64>(candidate_merged64, candidate_sorted32);
    TMRGSORT<256>(candidate_merged256, candidate_merged64);
    TMRGSORT<1024>(candidate_merged1024, candidate_merged256);

    FinalTopPairArray final_pair_blocks = TPARTVIEW<FinalTopPairTile, 1, CandidateCount / TopK>(candidate_merged1024);
    FinalTopPairView final_top_pairs = final_pair_blocks[0][0];
    FinalTopKValueTile topk_values;
    FinalTopKPositionTile selected_candidate_positions;
    FinalTopKIndexTile topk_indices;

    TGATHER(topk_values, final_top_pairs, MaskPattern::P0101);
    TGATHER(selected_candidate_positions, final_top_pairs, MaskPattern::P1010);
    TGATHER(topk_indices, candidate_indices, selected_candidate_positions);

    TopKResult result;
    result.Values = topk_values;
    result.Indices = topk_indices;
    return result;
}
```

### Required API/compiler closure

This code is proposed and not supported by the current headers. It depends on:

1. `TPARTVIEW : Tile -> BorrowedTileArray<SubTile,R,C>`.
2. `TASSEMBLY : TileArray<SubTile,R,C> -> Tile`.
3. Pair-aware `TSORT32(dst_pairs, values, absolute_indices)`.
4. Pair-aware `TMRGSORT<BlockLen>` preserving value/index association.
5. Mask-pattern gather with P0101 value lanes and P1010 index lanes.
6. TileArray destination slots accepted by TGATHER and lowered to the producing binder plus B.ASSEMBLE.
7. Final candidate position -> original ID gather through the separately assembled candidate-index parent.

Current Linx `TSORT(dst, src)` is not equivalent to PyPTO `sort32(values, indices)` because it has no explicit index input or typed pair output.

Compiler-side partition/session lowering is tracked by [LinxISA/llvm-project#74](https://github.com/LinxISA/llvm-project/issues/74). Programming-model review is [LinxISA/Linx-TileOP-API#36](https://github.com/LinxISA/Linx-TileOP-API/issues/36). Architecture/compiler contract is [LinxISA/linx-isa#187](https://github.com/LinxISA/linx-isa/pull/187).

### Repository routing

This standalone issue is a review/import signal. The maintained implementation and validation target remains the canonical nested source:

```text
workloads/pto_kernels/benchmarks/supernpu/benchmark/one-level-arch/kernels/sort/
```

Acceptance evidence must come from the nested pto-kernels tree with exact TileOP/LLVM/PTO/QEMU/model SHAs; standalone SuperNPUBench results alone are not promotion evidence.

**Comment by jiale-wangOwO** [comment]
Release-ops evidence from LinxISA/SuperScalarModel#641: the shipped one-level `topk.elf` is currently not an executable TopK implementation under Linx.

Source `benchmark/one-level-arch/test/kernel/sort/topk/topk.cpp` compiles both histogram launches only under `#ifndef __linx`; the `#else` branch explicitly states that the histogram stays zero. Therefore the Linx ELF retains only zero `TEXPANDS` plus `TSTORE` buffers, not the high8/low8 histogram computation.

Current single-PE gfrun reproduction of artifact SHA256 `0a812cc3ce894641e563e7b948a9311025ef7220c111fbb667e9030a2b088d00` observes zero high/low histograms, `kth_bin=0`, and `Match: 0/2048, R2=1`, exactly as that fallback predicts. This is not a gfrun TEXPANDS/TSTORE defect.

Please treat a Linx-equivalent histogram/extraction path and a newly generated ELF with independent numerical golden as required acceptance evidence for the topk work tracked here.
